mlflow/data.py

```diff
@@ -27,33 +33,30 @@ def _fetch_dbfs(uri, local_path):
 def _fetch_s3(uri, local_path):
     import boto3
     print("=== Downloading S3 object %s to local path %s ===" % (uri, os.path.abspath(local_path)))
-    (bucket, s3_path) = parse_s3_uri(uri)
+    (bucket, s3_path) = parse_simple_uri(uri, ["s3"])
     boto3.client('s3').download_file(bucket, s3_path, local_path)
 
 
 def _fetch_gs(uri, local_path):
     from google.cloud import storage
     print("=== Downloading GCS file %s to local path %s ===" % (uri, os.path.abspath(local_path)))
-    (bucket, gs_path) = parse_gs_uri(uri)
+    (bucket, gs_path) = parse_simple_uri(uri, ["gs"])
     storage.Client().bucket(bucket).blob(gs_path).download_to_filename(local_path)
 
 
-def parse_s3_uri(uri):
-    """Parse an S3 URI, returning (bucket, path)"""
-    parsed = urllib.parse.urlparse(uri)
-    if parsed.scheme != "s3":
-        raise Exception("Not an S3 URI: %s" % uri)
-    path = parsed.path
-    if path.startswith('/'):
-        path = path[1:]
-    return parsed.netloc, path
+def _fetch_hdfs(uri, local_path):
+    print("=== Downloading HDFS file %s to local path %s ===" % (uri, os.path.abspath(local_path)))
+    parse_simple_uri(uri, ["hdfs", "viewfs"])
+    host, port, path = _resolve_connection_params(uri)
```

Comment:
This uses private functions from `hdfs_artifact_repo`. Can we rewrite this use `HdfsArtifactRepository` directly?

Reply:
I made a change to use HdfsArtifactRepository directly, but I had to use a private method of the class that I implemented: _download_file.

I'm not sure it is what you meant with your comment, but I couldn't use the already existing method download_artifacts since it always create a temporary directory where the file will be downloaded, so the file wouldn't be downloaded in the output_path provided in data.py.

What I could do instead of this is that the method download_uri in data.py returns a new output path that we could get in _project_spec.py, so I could call download_artifacts and return the new path.